apps/studio/src/server/modules/user/inactiveUsers.service.ts

```diff
@@ -0,0 +1,241 @@
+    .where((eb) =>
+      eb.or([
+        // Users who have never logged in
+        eb.and([
+          eb("User.lastLoginAt", "is", null),
+          eb("User.createdAt", "<=", toDateThreshold),
+          ...(fromDateThreshold
+            ? [eb("User.createdAt", ">", fromDateThreshold)]
+            : []),
+        ]),
+        // Users who have logged in but haven't logged in for a while
+        eb.and([
+          eb("User.lastLoginAt", "is not", null),
+          eb("User.lastLoginAt", "<=", toDateThreshold),
+          ...(fromDateThreshold
+            ? [eb("User.lastLoginAt", ">", fromDateThreshold)]
+            : []),
+        ]),
+      ]),
+    )
+    .selectAll(["User"])
+    .distinct()
+    .execute()
+}
+
+interface DeactivateUsersProps {
+  tx: Transaction<DB>
+  userIds: User["id"][]
+}
+type DeactivateUsersOutput = {
+  user: User
+  siteIds: Site["id"][]
+}[]
+
+export const deactivateUsers = async ({
+  tx,
+  userIds,
+}: DeactivateUsersProps): Promise<DeactivateUsersOutput> => {
+  // prevent empty array from being passed in
+  if (userIds.length === 0) return []
+
+  const deletedPermissions = await tx
+    .updateTable("ResourcePermission")
+    .where("userId", "in", userIds)
+    .where("deletedAt", "is", null)
+    .set({ deletedAt: new Date() })
+    .returningAll()
+    .execute()
+
+  return tx
+    .selectFrom("User")
+    .where("id", "in", userIds)
+    .selectAll()
+    .execute()
+    .then((users) =>
+      users.map((user) => ({
+        user,
+        siteIds: deletedPermissions
+          .filter((permission) => permission.userId === user.id)
+          .map((permission) => permission.siteId),
+      })),
+    )
+}
+
+interface GetSiteAndAdminsProps {
+  tx: Transaction<DB>
+  userId: User["id"]
+  siteIds: Site["id"][]
+}
+
+type GetSiteAndAdminsOutput =
+  AccountDeactivationEmailTemplateData["sitesAndAdmins"]
+
+const getSiteAndAdmins = async ({
+  tx,
+  userId,
+  siteIds,
+}: GetSiteAndAdminsProps): Promise<GetSiteAndAdminsOutput> => {
+  // Prevent empty array from being passed in
+  if (siteIds.length === 0) return []
+
+  return (
+    tx
+      .with("siteAdmins", (eb) =>
+        eb
+          .selectFrom("Site")
+          .innerJoin(
+            "ResourcePermission",
+            "ResourcePermission.siteId",
+            "Site.id",
+          )
+          .innerJoin("User", "User.id", "ResourcePermission.userId")
+          .where("Site.id", "in", siteIds)
+          .where("ResourcePermission.userId", "!=", userId) // don't want to ask users to ask themselves for permissions
+          .where("ResourcePermission.deletedAt", "is", null)
+          .where("ResourcePermission.role", "=", RoleType.Admin) // should only give the admin emails to request reactivation permissions from
+          .where("User.email", "not in", ISOMER_ADMINS_AND_MIGRATORS_EMAILS) // we don't want to send emails to admins and migrators
```

Comment:
[side observation] this `ISOMER_ADMINS_AND_MIGRATORS_EMAILS` pattern comes up enough that we might benefit just from making a super admin table lol